osu.Server.Spectator/Hubs/ScoreUploader.cs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Game.Scoring;
+using osu.Server.Spectator.Database;
+using osu.Server.Spectator.Entities;
+using osu.Server.Spectator.Storage;
+using Timer = System.Timers.Timer;
+
+namespace osu.Server.Spectator.Hubs
+{
+    public class ScoreUploader : IEntityStore, IDisposable
+    {
+        /// <summary>
+        /// Amount of time (in milliseconds) between checks for pending score uploads.
+        /// </summary>
+        public double UploadInterval
+        {
+            get => timer.Interval;
+            set => timer.Interval = value;
+        }
+
+        /// <summary>
+        /// Amount of time (in milliseconds) before any individual score times out if a score ID hasn't been set.
+        /// This can happen if the user forcefully terminated the game before the API score submission request is sent, but after EndPlaySession() has been invoked.
+        /// </summary>
+        public double TimeoutInterval = 30000;
+
+        private bool shouldSaveReplays => Environment.GetEnvironmentVariable("SAVE_REPLAYS") == "1";
+
+        private readonly ConcurrentQueue<UploadItem> queue = new ConcurrentQueue<UploadItem>();
+        private readonly IDatabaseFactory databaseFactory;
+        private readonly IScoreStorage scoreStorage;
+        private readonly Timer timer;
+        private readonly CancellationTokenSource timerCancellationSource;
+        private readonly CancellationToken timerCancellationToken;
+
+        public ScoreUploader(IDatabaseFactory databaseFactory, IScoreStorage scoreStorage)
+        {
+            this.databaseFactory = databaseFactory;
+            this.scoreStorage = scoreStorage;
+
+            timerCancellationSource = new CancellationTokenSource();
+            timerCancellationToken = timerCancellationSource.Token;
+
+            timer = new Timer(5000);
```

Comment:
I think this needs to run more like every 50 or 100 ms. We'd want scores to be uploaded ASAP, else users may visit the score page and have no replay available for download.